### PR TITLE
OPAL: fix string buffer allocation for large env variables

### PR DESCRIPTION
**Pull request by devreal**
In OPAL's keyval parser in `add_to_env_str`, for the first variable a buffer of size 1024 was allocated, regardless of the space needed. For any additional variables, the size might have been doubled at best, even if more space was required. This PR makes sure the allocated buffer is sufficiently large to store the value.

See https://github.com/open-mpi/ompi/issues/8117

*Note:* the issue described in #8117 does not seem to occur on `master`. I will backport this to 4.0.x (where it occurs) and 4.1.x (not tested yet) but the same code exists in PMIX and PRRTE so it should be picked up there as well.

Signed-off-by: Joseph Schuchart <schuchart@icl.utk.edu>

### CODE REVIEW

**Review by bosilca — APPROVED** [review]

File: opal/util/keyval_parse.c [hunk 1]
```diff
@@ -276,33 +276,38 @@ static int save_param_name (void)
...
-    if (NULL != env_str) {
-        varsz = strlen(var);
-        valsz = (NULL != val) ? strlen(val) : 0;
-        sz = strlen(env_str)+varsz+valsz+2;
-        if (envsize <= sz) {
-            envsize *=2;
+    varsz = strlen(var);
+    valsz = (NULL != val) ? strlen(val) : 0;
+    sz = (NULL != env_str) ? strlen(env_str) : 0;
+    sz += varsz+valsz+2;
```
Comment (bosilca) [inline opal/util/keyval_parse.c:289]: If you want to be picky that's one wasted byte when ```NULL != env_str```.

**Comment by jsquyres** [comment]
FYI @rhc54 @jjhursey for the "we should also do this in PMIX and PRRTE" note.

File: opal/util/keyval_parse.c [hunk 2]
```diff
@@ -276,33 +276,38 @@ static int save_param_name (void)
...
-            tmp = realloc(env_str, envsize);
+    if (NULL != env_str) {
+        if (new_envsize > envsize) {
+            tmp = realloc(env_str, new_envsize);
             if (NULL == tmp) {
                 return OPAL_ERR_OUT_OF_RESOURCE;
             }
             env_str = tmp;
         }
         strcat(env_str, ";");
```
Comment (jsquyres) [inline opal/util/keyval_parse.c:303]: Is there a `+1` missing somewhere to account for the `;` character?

Comment (jsquyres) [reply opal/util/keyval_parse.c:303]: There's also a `strcat()` of an additional `=` character on old line 309.  Does that need to be counted, too?

Comment (devreal) [reply opal/util/keyval_parse.c:315]: You're right, the original version didn't account for the terminating NUL byte. The chance of hitting that is pretty slim as allocations are made in multiples of 1K but I force-pushed an update  that makes accounting =, ;, and \0 explicit.

**Review by jsquyres — APPROVED** [review]

**Comment by jsquyres** [comment]
Please also update #8155 and #8156.  Thanks!

**Pull request merged**